### PR TITLE
Add a TychoReactorReader that mimics maven behavior for Tycho packagings

### PR DESCRIPTION
**Pull request by laeubi**
FYI @lppedd lets see if this makes anything worse ...

This includes 
- https://github.com/eclipse-tycho/tycho/pull/1407

### CODE REVIEW

File: tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    private boolean isTychoReactorArtifact(Artifact artifact) {
+        if (artifact.getClassifier() == null || artifact.getClassifier().isBlank()) {
+            return PackagingType.TYCHO_PACKAGING_TYPES.contains(artifact.getProperty("type", ""));
+        }
+        return false;
+    }
+
+    @Override
+    public List<String> findVersions(Artifact artifact) {
+        return getTychoReactorProject(artifact).map(project -> List.of(artifact.getVersion())).orElse(List.of());
```
Comment (lppedd) [inline tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java:93]: In my custom extension I always return an empty list here. 
If I recall correctly this is basically unused.

Comment (laeubi) [reply tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java:93]: `DefaultVersionRangeResolver` calls this method.

File: tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @Override
+    public File findArtifact(Artifact artifact) {
+        return getTychoReactorProject(artifact).map(project -> {
+            org.apache.maven.artifact.Artifact mainArtifact = project.getArtifact();
+            if (mainArtifact != null && mainArtifact.getFile() != null && mainArtifact.getFile().exists()) {
+                return mainArtifact.getFile();
+            }
+            return project.getBasedir();
```
Comment (lppedd) [inline tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java:66]: Not sure if it's more correct to use `project.getBasedir()` or `project.getBuild().getOutputDirectory()` (latter is what I currently return).

`ReactorReader` does this:
```java
if (isTestArtifact( artifact )) {
    if (project.hasLifecyclePhase("test-compile")) {
        return new File(project.getBuild().getTestOutputDirectory());
    }
} else {
    String type = artifact.getProperty("type", "");
    if (project.hasLifecyclePhase("compile") && COMPILE_PHASE_TYPES.contains(type)) {
        return new File(project.getBuild().getOutputDirectory());
    }
}
```
And that's why I've picked `project.getBuild().getOutputDirectory()`.

**Comment by lppedd** [comment]
I can definitely say this is going to work, as I already use this solution!

File: tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java (hunk 1 above)
Comment (lppedd) [reply tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java:93]: @laeubi ahh yes, you're right, not sure why the search didn't bring up any reference.

File: tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java (hunk 2 above)
Comment (laeubi) [reply tycho-core/src/main/java/org/eclipse/tycho/core/maven/TychoReactorReader.java:66]: Tycho uses the basedir in such situations already so this is consistent (even though a bit bad and we might later want to get rid of it).

**Comment by laeubi** [comment]
> I can definitely say this is going to work, as I already use this solution!

It seems to pass the testsuite and your new test is also recognized, but lets wait for the Jenkins build ... the problem is that there are sometimes quite special cases in Tycho so even if it works most of the times one can be surprised some times :-)

File: tycho-its/src/test/java/org/eclipse/tycho/test/resolver/TychoGenerateSourcesIssueTest.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import org.apache.maven.it.Verifier;
+import org.eclipse.tycho.test.AbstractTychoIntegrationTest;
+import org.junit.Test;
+
+public class TychoGenerateSourcesIssueTest extends AbstractTychoIntegrationTest {
+    @Test
+    public void testTychoGenerateSourcesIssue() throws Exception {
+        Verifier verifier = getVerifier("resolver.generateSources", false, false);
+        verifier.addCliOption("-Dtycho.resolver.classic=false");
```
Comment (lppedd) [inline tycho-its/src/test/java/org/eclipse/tycho/test/resolver/TychoGenerateSourcesIssueTest.java:13]: Just noticed this. Maybe it's better to remove this line?

**Pull request merged**